Update XRTMap to handle the missing metadata for the observer

nabobalis opened the pull request:
Fixes https://github.com/sunpy/sunpy/issues/6315 as well

TODO:
- [x] Changelog

nabobalis commented on sunpy/map/sources/tests/test_xrt_source.py:
```diff
@@ -13,45 +13,49 @@
     # Smoke test that WCS is valid and can transform from pixels to world coordinates
-    with pytest.warns(SunpyMetadataWarning, match='Missing metadata for observer'):
-        xrt.pixel_to_world(0*u.pix, 0*u.pix)
+    with pytest.warns(SunpyMetadataWarning, match='the following metadata is missing: dsun_obs'):
```
Level 0 files miss this but the Level 1 files do have it.

nabobalis commented on sunpy/map/sources/hinode.py:
```diff
@@ -36,6 +36,9 @@ class XRTMap(GenericMap):
                                   "Be_thick", "Gband", "Ti_poly"]
 
     def __init__(self, data, header, **kwargs):
+        # https://community.openastronomy.org/t/sunpymetadatawarnings-when-using-hinode-xrt-data/393/7
+        header["HGLN_OBS"] = 0
+        header["HGLT_OBS"] = 0
```
I could add more of a comment but I feel like the link is good enough?

dstansby reviewed (changes requested):
Since https://github.com/sunpy/sunpy/pull/5501 our approach to source specific fixes has been to override map properties, and not to modify metadata, so this PR should be updated to take that approach (presumably to override `.observer_coordinate` property, but that's off the top of my head without checking) .

dstansby commented:
I am also of the view that whilst this might be a fine approach to take, we should:
1. Quantify and document how big the potential error is in setting these values to zero (probably via asking on the OpenAstronomy thread)
2. Possibly still emit some warning (but a single one that's source specific instead of the current ones) because we are setting innacurate data that could trip a user up depending on how accurate they need the observer coordinate information for their analysis.

Cadair commented:
I think you will want to override the property below `.observer_coordinate` the one that reads the groups of keys out of the header.

nabobalis commented:
I think I override it correctly?

dstansby commented on sunpy/map/sources/hinode.py:
```diff
@@ -1,8 +1,8 @@
+import astropy.units as u
+from astropy.coordinates import Latitude, Longitude
 
-__author__ = ["Jack Ireland, Jose Ivan Campos-Rozo, David Perez-Suarez"]
-__email__ = "jack.ireland@nasa.gov"
```
Should probably not remove this until we resolve https://github.com/sunpy/sunpy/issues/3650

nabobalis replied:
I think it should really go and we should decide on #3650.

dstansby replied:
I agree, but given there's still disagreement lets resolve this in a different PR.

nabobalis replied:
If we don't decide tomorrow about this, I will undo these changes.

dstansby reviewed (changes requested):
- I think you're going to need to override `.observer_coordinate` too to prevent the warning.
- I'm stil +0.5 on emitting some sort of warning here, but there should defintitely be a clear explanation of what we're assuming and why, and this is where a link to the forum post can go instead of being hidden in the comments
- We need to allow for the case where the metadata is present in the files and not assume 0 lat/lon in that case, e.g. if a user has manually set these values in the files before loading them. This will need a test too.

ayshih reviewed (changes requested):
- You'll need to override `.observer_coordinate` property (and not the `.heliographic_latitude`/`.heliographic_longitude` properties).  I don't think there's a way to fix this by overriding `._supported_observer_coordinates` property, because there is no keyword to use for the heliographic longitude.  The overridden `.observer_coordinate` property should check if it needs to apply the assumption, since there may be user-added observer metadata.
- The heliographic latitude should use the value in `SOLAR_B0`.  It should not be set to 0 degrees.
- This assumption should be stated in the `XRTMap` docstring.  I don't think it's necessary to quantify the uncertainty.  I also don't think any warning should be emitted.

ayshih commented:
Hinode is in low-Earth orbit (LEO) with an altitude of ~600 km.  Thus, it is always <1.1 Earth radii from Earth center.  (For comparison, SDO is ~6.5 Earth radii from Earth center.)  This amount of offset is negligible for normal imaging of the Sun.  The exception cases are when there is an intermediate body – Venus or the Moon – occulting part of the Sun.

nabobalis commented:
Ok, I updated it and rebased the PR.

I still think I did it wrong especially when it comes to settings the radius in the creation of the new skycoord.

The code isn't pretty either and I need to add a test to make sure if someone provides the correct metadata, it will not be overridden.

nabobalis commented on sunpy/map/sources/hinode.py:
```diff
@@ -36,24 +46,43 @@ class XRTMap(GenericMap):
-        super().__init__(data, header, **kwargs)
-
-        fw1 = header.get('EC_FW1_')
+        fw1 = header.get('EC_FW1_', '')
```
Maybe we want this code to error if the keyword is missing wholesale?

nabobalis commented on sunpy/map/sources/hinode.py:
```diff
@@ -36,24 +46,43 @@ class XRTMap(GenericMap):
+        return super()._timesys
+
+    @property
+    # @cached_property_based_on('_meta_hash')
```
I was unsure if I should restore this behaviour here.

dstansby replied:
I can't see why not?

nabobalis commented on sunpy/map/sources/hinode.py:
```diff
@@ -21,6 +24,13 @@ class XRTMap(GenericMap):
 
+    .. note::
+
+        If the required parameters for observator are not provided,
```
TOOD: Check this renders nicely.

dstansby replied:
You should be able to use
```
Notes
-----
```
To create a notes heading

ayshih commented:
> I don't think there's a way to fix this by overriding `._supported_observer_coordinates` property, because there is no keyword to use for the heliographic longitude.

Whoops, I refreshed myself on how `._supported_observer_coordinates` actually works, and you should override that instead of `.observer_coordinate`.  I'll put up an (untested) code suggestion.

ayshih commented on sunpy/map/sources/hinode.py:
```diff
@@ -36,24 +46,43 @@ class XRTMap(GenericMap):
+        return super()._timesys
+
+    @property
+    # @cached_property_based_on('_meta_hash')
+    def observer_coordinate(self):
+        # Based on the discussion from
+        # # https://community.openastronomy.org/t/sunpymetadatawarnings-when-using-hinode-xrt-data/393/7
+        # Certain keywords are missing from the level 1 header that make haev to set values
+        # to avoid warnings being raised.
+        missing_meta = {}
+        for keys, kwargs in self._supported_observer_coordinates:
+            if not isinstance(kwargs['frame'], str):
+                kwargs['frame'] = kwargs['frame'].name
+            missing_meta[kwargs['frame']] = set(keys).difference(self.meta.keys())
+            if "heliographic_stonyhurst" in missing_meta:
+                if "hgln_obs" in missing_meta["heliographic_stonyhurst"]:
+                    extra_kwargs = {"lon": 0,
+                                    "lat": self.meta.get("SOLAR_B0", sun.B0(self.meta.get("DATE_OBS")).to_value(u.deg)),
+                                    "radius": constants.radius
+                                    }
+                    sc = SkyCoord(obstime=self.date, **{**kwargs, **extra_kwargs})
+                    sc = sc.heliographic_stonyhurst
+                    return SkyCoord(sc.replicate(rsun=self._rsun_meters(sc.radius)))
+        return super().observer_coordinate
```
```suggestion
    @property
    def _supported_observer_coordinates(self):
        return (super()._supported_observer_coordinates
                + [(('solar_b0', 'dsun_obs'), {'lon': 0*u.deg,
                                               'lat': self.meta.get('solar_b0'),
                                               'radius': self.meta.get('dsun_obs'),
                                               'unit': (u.deg, u.deg, u.m),
                                               'frame': "heliographic_stonyhurst"})])
```
Note that unlike our other map sources, we add this option with the assumption at the end of the list so that we'll automatically use observer metadata if it already exists.

ayshih commented on sunpy/map/sources/hinode.py:
```diff
@@ -21,6 +24,13 @@ class XRTMap(GenericMap):
 
     Hinode was launched on 22 September 2006 into a sun-synchronous orbit.
 
+    .. note::
+
+        If the required parameters for observator are not provided,
+        default values will be used for ``HGLT_OBS`` and ``HGLN_OBS``.
+        ``HGLN_OBS`` will be set to 0 degrees.
+        ``HGLT_OBS`` will be set to the B0 value.
```
```suggestion
    Notes
    -----
    XRT files do not normally specify the heliographic longitude of the spacecraft,
    and the default assumption is that the spacecraft is at zero Stonyhurst
    heliographic longitude (i.e., the same longitude as Earth).  This assumption is
    safe for nearly all analyses due to Hinode's orbital altitude of only ~600 km.
    This assumption is not made if ``HGLN_OBS`` and ``HGLT_OBS`` values have been
    explicitly added to the metadata.
``````

nabobalis commented on sunpy/map/mapbase.py:
```diff
@@ -1213,7 +1213,7 @@ def observer_coordinate(self):
                 return SkyCoord(sc.replicate(rsun=self._rsun_meters(sc.radius)))
 
-            elif any(meta_list) and not set(keys).isdisjoint(self.meta.keys()):
+            elif not any(meta_list) and set(keys).isdisjoint(self.meta.keys()):
```
I think this is correct?

nabobalis replied:
Clearly not.

nabobalis replied:
I think this line should be dropped and we work out the missing keys all time instead?

nabobalis replied:
Tweaked now.

ayshih commented:
(or add a newline to the end of the warning message)

nabobalis commented:
Will do, once I fix my broken local env, I will make sure it works. This time.

nabobalis commented on sunpy/data/_sample.py:
```diff
@@ -93,7 +93,7 @@ def _retry_sample_data(results, new_url_base):
 
     for err in results.errors:
-        file_name = err.filepath_partial().name
+        file_name = err.url.split("/")[-1]
```
This will probably conflict with the open PR for parfive.

dstansby commented on sunpy/map/sources/tests/test_xrt_source.py:
```diff
@@ -13,45 +13,58 @@
+
+
+def test_heliographic_longitude(xrt_map):
+    with pytest.warns(SunpyMetadataWarning, match='Missing metadata for observer'):
```
I thought the point was to get rid of these warnings?

nabobalis replied:
We use a level 0 header from what I can see and that will raise this warning. 

In a level 1 header (which I still need to source), it should not raise this warning.

dstansby replied:
Are we not trying to get rid of the warning for the L0 header too?

nabobalis replied:
I think we would need to calculate solar_b0 if its missing for the level 0 headers but I think people are meant to use the level 1 files.

ayshih replied:
I am slightly opposed to suppressing the warning for L0 files

dstansby replied:
Why? Doesn't the same logic for suppressing the warning for L1 files apply for L0 files?

ayshih replied:
Here's the distinction that I'm drawing.  When generating the L1 files, the XRT team has decided that specifying `SOLAR_B0` and `DSUN_OBS` is sufficient observer metadata, and that the omission of `HGLN_OBS` shall be interpreted as `HGLN_OBS = 0`.  This PR implements that specific handling of an intentionally omitted keyword.  In contrast, the L0 files are lacking observer metadata entirely, so to me are still a bit raw rather than "analysis-ready".  Since our assuming that the observer is at Earth center injects information rather than fixes how we interpret existing information, it feels more dramatic and still deserving of a warning, even if the answer turns out to be the same as what the XRT team does when promoting L0 to L1.

dstansby commented on sunpy/map/sources/hinode.py:
```diff
@@ -36,24 +47,31 @@ class XRTMap(GenericMap):
+    def _supported_observer_coordinates(self):
+        # Assume observer is at zero Stonyhurst heliographic longitude if not otherwise specified
+        # https://community.openastronomy.org/t/sunpymetadatawarnings-when-using-hinode-xrt-data/393/7 for more information
+        return (super()._supported_observer_coordinates
+                + [(('solar_b0', 'dsun_obs'), {'lon': 0*u.deg,
+                                               'lat': self.meta.get('solar_b0'),
+                                               'radius': self.meta.get('dsun_obs'),
+                                               'unit': (u.deg, u.deg, u.m),
+                                               'frame': "heliographic_stonyhurst"})])
```
Is `'solar_b0'` in the meadata, so we don't actually need to calculate it ourselves?

nabobalis replied:
In the level 1 data it should be, we don't have a level 1 header at the moment.

dstansby reviewed (changes requested):
Wasn't the point of this PR to get rid of the warnings? If so there shouldn't be an `pytest.warns` for missing metadata in the tests any more.

This should also get a new test to check that manually entered metadata takes preference over the new assumption.

Otherwise looks 👍

dstansby commented on changelog/6333.bugfix.rst:
```diff
@@ -0,0 +1,3 @@
+If an XRT file does not specify the heliographic longitude of the spacecraft,
```
```suggestion
If a level 1 XRT file does not specify the heliographic longitude of the spacecraft,
```

dstansby commented on changelog/6333.bugfix.rst:
```diff
@@ -0,0 +1,3 @@
+If an XRT file does not specify the heliographic longitude of the spacecraft,
+the default assumption is that the spacecraft is at zero Stonyhurst
+heliographic longitude (i.e., the same longitude as Earth).
```
```suggestion
heliographic longitude (i.e., the same longitude as Earth) is silently assumed.
```

dstansby commented on sunpy/map/sources/hinode.py:
```diff
@@ -21,6 +23,15 @@ class XRTMap(GenericMap):
+    Notes
+    -----
+    XRT files do not normally specify the heliographic longitude of the spacecraft,
+    and the default assumption is that the spacecraft is at zero Stonyhurst
+    heliographic longitude (i.e., the same longitude as Earth). This assumption is
```
```suggestion
    so sunpy silently assumes is that the spacecraft is at zero Stonyhurst
    heliographic longitude (i.e., the same longitude as Earth) for L1 files. This assumption is
```

dstansby reviewed (changes requested):
If I'm correct we haven't added a test for this - I think we should add a L1 header and test map, and add a test to
- check the warning isn't being raised.
- check that if the longitude is in the file, that value is used instead of the deafult 0 deg value we've added

See also my suggestions to make the docstring/changelog clearer.

nabobalis commented:
Updated the test header to a Level 1 header and we don't have any errors raised with it as things stand with this PR.

dstansby commented on changelog/6333.bugfix.rst:
```diff
@@ -0,0 +1,3 @@
+If a level 1 XRT file does not specify the heliographic longitude of the spacecraft,
+the default assumption is that the spacecraft is at zero Stonyhurst
+heliographic longitude (i.e., the same longitude as Earth) is silently assumed.
```
```suggestion
If a level 1 XRT file does not specify the heliographic longitude of the spacecraft,
a silent assumption is made that the spacecraft is at zero Stonyhurst
heliographic longitude (i.e., the same longitude as Earth).
```

ayshih commented on changelog/6333.bugfix.rst:
```diff
@@ -0,0 +1,3 @@
+If a level 1 XRT file does not specify the heliographic longitude of the spacecraft,
+a silent assumption is made that the spacecraft is at zero Stonyhurst
+heliographic longitude (i.e., the same longitude as Earth) is silently assumed.
```
```suggestion
heliographic longitude (i.e., the same longitude as Earth).
```
Heh, I don't know what you guys are doing, but https://github.com/sunpy/sunpy/pull/6328/commits/df075b73b5c1b373423bbf70777b0176546030a1 should not have been committed. =P

ayshih commented on sunpy/map/sources/hinode.py:
```diff
@@ -21,6 +23,15 @@ class XRTMap(GenericMap):
+    Notes
+    -----
+    XRT files do not normally specify the heliographic longitude of the spacecraft,
+    so sunpy silently assumes is that the spacecraft is at zero Stonyhurst
+    heliographic longitude (i.e., the same longitude as Earth) for L1 files. This assumption is
```
```suggestion
    so sunpy silently assumes that the spacecraft is at zero Stonyhurst heliographic
    longitude (i.e., the same longitude as Earth) for L1 files. This assumption is
```

ayshih reviewed (approved):
Some docstring tweaks, otherwise looks good to me
